posts/ethereum-development-tips.md

```diff
@@ -0,0 +1,254 @@
+
+    uint256 i = 42;
+
+    address x =  0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2;
+
+Another fun thing to manage close to C is of course the storage. Objects do not always behave the way you want them to and you need to remember that everything has a price on a blockchain.
+
+You have to know if the object you are defining in your function needs to be “memory” (stored temporarily) or “storage” (knowing that it will be stored in the blockchain). For example, a dynamic array has to be storage, so you may need to find a way to know the exact number of object it should need.
+
+Other oddities may include structures not being able to be defined recursively, or more importantly, all state variables being initialized to their default value. This means that a boolean will be automatically false, and also that a mapping (Solidity’s widely used equivalent of a hashmap) will be initialized for every key at the byte 0. This means that you cannot iterate on the keys!
+
+    LinkedList.Data  storage list;
+
+    uint256[]  memory _parts;
+
+Complex storage structures can be useful, such as the Linked List we needed to implement. By using a mapping from a uint id to a Node structure (with the data we want, the information to it will be sorted by, and the id of the next Node) and keeping the id of the head of the list, we could make an easy sortable list where we can add and delete nodes.
+
+One of the burden of Solidity is how to code a secure smart contract. Writing a contract that behaves as you expect is quit easy, however making sure it won’t later act strangely is hader. Solidity, is a constantly evolving language, but can still fall short of being a secure language for creating decentralized application.
+
+One of the tools provided by Solidity is for Error handling. You will probably want to have quite a few “Require” functions used throughout your code, as they will “Revert” (which means cancelling the execution of the transaction and revert back any gas used to the sender) when you need them to. They can be used to check users balance before transfers for instance.
+
+Now let’s go over a few problems or considerations you may encounter while coding in Solidity (well, that we have encountered or seen in solidity hacking challenges). A lot of them are really specific to the way blockchain works.
+
+The most common and biggest problem to check out is the classic reentrancy problem. For example, you do not want a user to be able to call two times a function (or two different functions) that will give him tokens before depleting his balance!
```

Comment:
Not sure this is reentrancy. Reentrancy is the capacity of a function to call itself (re-enter).